src/group_theory/noncomm_pi_coprod.lean

```diff
@@ -0,0 +1,426 @@
+/-
+Copyright (c) 2022 Jocchim Breitner. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Joachim Breitner
+-/
+import group_theory.order_of_element
+import data.finset.noncomm_prod
+import data.fintype.card
+
+/-!
+# Canonical homomorphism from a finite family of monoids
+
+This file defines the construction of the canoncial homomorphism from a family of monoids.
+
+Given a family of morphisms `ϕ i : N i →* M` for each `i : ι` where elements in the
+image of different morphism commute, we obtain a canoncial morphism
+`monoid_hom.noncomm_pi_coprod : (Π i, N i) →* M` that coincides with `ϕ`
+
+## Main definitions
+
+* `monoid_hom.noncomm_pi_coprod : (Π i, N i) →* M` is the main homomorphism
+* `noncomm_pi_coprod_on.hom (S : finset S) : (Π (i : ι), N i) →* M` is the homomorphism
+   restricted to the set `S`, and mostly of internal interest in this file, to allow inductive
+   proofs (and thus in its own namespace).
+* `subgroup.noncomm_pi_coprod : (Π i, H i) →* G` is the specialization to `H i : subgroup G`
+   and the subgroup embedding.
+
+## Main theorems
+
+* `monoid_hom.noncomm_pi_coprod` conicides with `ϕ i` when restricted to `N i`
+* `monoid_hom.noncomm_pi_coprod_mrange`: The range of `monoid_hom.noncomm_pi_coprod` is
+  `⨆ (i : ι), (ϕ i).mrange`
+* `monoid_hom.noncomm_pi_coprod_range`: The range of `monoid_hom.noncomm_pi_coprod` is
+  `⨆ (i : ι), (ϕ i).range`
+* `subgroup.noncomm_pi_coprod_range`: The range of `subgroup.noncomm_pi_coprod` is `⨆ (i : ι), H i`.
+* `monoid_hom.injective_noncomm_pi_coprod_of_independent`: in the case of groups, `pi_hom.hom` is
+   injective if the `ϕ` are injective and the ranges of the `ϕ` are independent.
+* `monoid_hom.independent_range_of_coprime_order`: If the `N i` have coprime orders, then the ranges
+   of the `ϕ` are independent.
+* `subgroup.independent_of_coprime_order`: If commuting, normal subgroups `H i` have coprime orders,
+   they are independent.
+
+-/
+
+open_locale big_operators
+
+lemma coprime_prod_left
+  {ι : Type*}
+  {x : ℕ} {s : ι → ℕ} {t : finset ι} :
+  (∀ (i : ι), i ∈ t → nat.coprime (s i) x) → nat.coprime (∏ (i : ι) in t, s i) x :=
+finset.prod_induction s (λ y, y.coprime x) (λ a b, nat.coprime.mul) (by simp)
```

Comment:
Can you add this for `prod_right` too? (and find an appropriate file for both)

Up to you if you want to do it as a separate PR - although that will reduce your build time if this one undergoes a few more rounds of review.

Reply:
https://github.com/leanprover-community/mathlib/pull/12268